### PR TITLE
sys_storage vsh implementation

### PR DESCRIPTION
**Pull request by clienthax**

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 1]
```diff
@@ -1,37 +1,85 @@
...
 	sys_storage.todo("sys_storage_open(device=0x%x, mode=0x%x, fd=*0x%x, flags=0x%x)", device, mode, fd, flags);
 
-	return CELL_OK;
+	if (device == 0)
+	{
+		return CELL_ENOENT;
+	}
+
+	u64 storage_id = device & 0xFFFFF00FFFFFFFF;
+	fs::file file;
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:25]: What is the meaning empty file here?

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 2]
```diff
@@ -1,37 +1,85 @@
...
+		return CELL_ESRCH; // idk
+	}
+
+	if (handle->device_id == 0x100000200000004 && handle->file)
+	{
+		handle->file.seek(start_sector * 0x200);
+		u64 size = num_sectors * 0x200;
+		const u64 result = handle->file.read(bounce_buf.get_ptr(), size);
+
+		if (result != size) // mjau
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:66]: meow

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_storage.cpp:25]: file is const as well, it remains empty yet other places suggest it should contain somthing.

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 3]
```diff
@@ -1,37 +1,85 @@
...
 
 #include "Emu/Cell/ErrorCodes.h"
+#include "Emu/Cell/lv2/sys_event.h"
 
 #include "sys_storage.h"
 
-
 LOG_CHANNEL(sys_storage);
 
+u32 asyncequeue;
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:13]: Use FXO.

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 4]
```diff
@@ -1,37 +1,85 @@
 #include "stdafx.h"
+#include "Emu/System.h"
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:2]: I don't think this include is necessary

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp (hunk 1 above)
Comment (clienthax) [reply rpcs3/Emu/Cell/lv2/sys_storage.cpp:25]: I believe there used to be code to load the nor region file in there.., it doesn't seem to effect anything however, and im not sure what it should look like myself.

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 5]
```diff
@@ -46,13 +91,22 @@ error_code sys_storage_async_configure(u32 fd, u32 io_buf, u32 equeue_id, u32 un
 {
 	sys_storage.todo("sys_storage_async_configure(fd=0x%x, io_buf=0x%x, equeue_id=0x%x, unk=*0x%x)", fd, io_buf, equeue_id, unk);
 
+	auto& manager = g_fxo->get<storage_manager>();
+	manager.asyncequeue = equeue_id;
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:95]: Use the usual idm::get<lv2_obj, lv2_event_queue>(equeue_id) with ESRCH should suffuc, the type of storage_manager.asyncqueue should be atomic_ptr<std::shared_ptr<lv2_event_queue>>.

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 6]
```diff
@@ -81,20 +135,207 @@ error_code sys_storage_get_device_info(u64 device, vm::ptr<StorageDeviceInfo> bu
...
+	*devices = 17;
+
 	return CELL_OK;
 }
 
 error_code sys_storage_report_devices(u32 storages, u32 start, u32 devices, vm::ptr<u64> device_ids)
 {
 	sys_storage.todo("sys_storage_report_devices(storages=0x%x, start=0x%x, devices=0x%x, device_ids=0x%x)", storages, start, devices, device_ids);
 
+	std::array<u64, 0x11> all_devs;
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:312]: This seems like it should a static constexpr initialized by an immediate lambda.

File: rpcs3/Emu/Cell/lv2/sys_storage.h [hunk 7]
```diff
@@ -3,6 +3,49 @@
 #include "Emu/Memory/vm_ptr.h"
 #include "Emu/Cell/ErrorCodes.h"
 
+struct storage_manager
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_storage.h:6]: Definition in cpp file is better, also under annonymouse namespace.

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 8]
```diff
@@ -1,37 +1,89 @@
...
+	{
+		return CELL_ENOENT;
+	}
+
+	u64 storage_id = device & 0xFFFFF00FFFFFFFF;
+	fs::file file;
+
+	if (const u32 id = idm::make<lv2_storage>(device, std::move(file), mode, flags))
+	{
+		*fd = id;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:36]: fd might be null

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 9]
```diff
@@ -1,37 +1,89 @@
...
+	idm::remove<lv2_storage>(fd);
+
 	return CELL_OK;
 }
 
 error_code sys_storage_read(u32 fd, u32 mode, u32 start_sector, u32 num_sectors, vm::ptr<void> bounce_buf, vm::ptr<u32> sectors_read, u64 flags)
 {
 	sys_storage.todo("sys_storage_read(fd=0x%x, mode=0x%x, start_sector=0x%x, num_sectors=0x%x, bounce_buf=*0x%x, sectors_read=*0x%x, flags=0x%x)", fd, mode, start_sector, num_sectors, bounce_buf, sectors_read, flags);
 
+	memset(bounce_buf.get_ptr(), 0, num_sectors * 0x200);
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:56]: bounce_buf might be null

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 10]
```diff
@@ -1,37 +1,89 @@
...
+		u64 size = num_sectors * 0x200;
+		const u64 result = handle->file.read(bounce_buf.get_ptr(), size);
+
+		if (result != size) // mjau
+		{
+			fmt::throw_exception("didnt read expected");
+		}
+	}
+
+	*sectors_read = num_sectors;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:76]: sectors_read might be null

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 11]
```diff
@@ -1,37 +1,89 @@
...
+
 	return CELL_OK;
 }
 
 error_code sys_storage_write(u32 fd, u32 mode, u32 start_sector, u32 num_sectors, vm::ptr<void> data, vm::ptr<u32> sectors_wrote, u64 flags)
 {
 	sys_storage.todo("sys_storage_write(fd=0x%x, mode=0x%x, start_sector=0x%x, num_sectors=0x%x, data=*=0x%x, sectors_wrote=*0x%x, flags=0x%llx)", fd, mode, start_sector, num_sectors, data, sectors_wrote, flags);
 
+	*sectors_wrote = num_sectors;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:85]: sectors_wrote might be null

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 12]
```diff
@@ -81,20 +151,207 @@ error_code sys_storage_get_device_info(u64 device, vm::ptr<StorageDeviceInfo> bu
 {
 	sys_storage.todo("sys_storage_get_device_info(device=0x%x, buffer=*0x%x)", device, buffer);
 
+	memset(buffer.get_ptr(), 0, sizeof(StorageDeviceInfo));
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:154]: buffer might be null

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 13]
```diff
@@ -81,20 +151,207 @@ error_code sys_storage_get_device_info(u64 device, vm::ptr<StorageDeviceInfo> bu
...
+
 	return CELL_OK;
 }
 
 error_code sys_storage_get_device_config(vm::ptr<u32> storages, vm::ptr<u32> devices)
 {
 	sys_storage.todo("sys_storage_get_device_config(storages=*0x%x, devices=*0x%x)", storages, devices);
 
+	*storages = 6;
+	*devices = 17;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:319]: might be null

File: rpcs3/Emu/Cell/lv2/sys_storage.cpp [hunk 14]
```diff
@@ -81,20 +151,207 @@ error_code sys_storage_get_device_info(u64 device, vm::ptr<StorageDeviceInfo> bu
...
+	for (int i = 0; i < 4; ++i)
+	{
+		all_devs[i + 12] = 0x100000000000004 | (static_cast<u64>(i) << 32);
+	}
+
+	all_devs[16] = 0x100000000000003;
+
+	for (u32 i = 0; i < devices; ++i)
+	{
+		device_ids[i] = all_devs[start++];
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_storage.cpp:352]: device_ids might be null

**Pull request merged**